README.md
# multi-elo

[![npm version][npm-version-src]][npm-version-href]
[![npm downloads][npm-downloads-src]][npm-downloads-href]
[![bundle][bundle-src]][bundle-href]
[![Codecov][codecov-src]][codecov-href]

This package implements a multiplayer extension of the popular Elo rating system.

- [Installation](#installation)
- [Example Usage](#example-usage)
- [Methodology](#methodology)
  - [Traditional Elo ratings](#traditional-elo-ratings)
  - [Extension to multiplayer](#extension-to-multiplayer)

This Package is based on [djcunningham0s python implementation](https://github.com/djcunningham0/multielo/).
For additional information, see his [blog post](https://towardsdatascience.com/developing-a-generalized-elo-rating-system-for-multiplayer-games-b9b495e87802) on Towards Data Science (or try [this link](https://towardsdatascience.com/developing-a-generalized-elo-rating-system-for-multiplayer-games-b9b495e87802?sk=89615c121aa78c7b502e9dce35ece5e1) if you hit a paywall).

## Installation

The package can be installed from GitHub or npm by using `npm`.

```sh
# npm
npm install multi-elo

# yarn
yarn add multi-elo

# pnpm
pnpm install multi-elo

# bun
bun install multi-elo
```

Import:

```js
// ESM
import {} from "multi-elo";

// CommonJS
const {} = require("multi-elo");
```

## Example Usage

The following example shows how to calculate updated Elo ratings after a matchup using the default settings in the package.

```typescript
import { MultiElo } from 'multi-elo';

# player with 1200 rating beats a player with 1000 rating
MultiElo.getNewRatings([1200, 1000])
# [1207.68809835,  992.31190165]

# player with 900 rating beats player with 1000 rating
MultiElo.getNewRatings([900, 1000])
# [920.48207999, 979.51792001]

# 3-way matchup
MultiElo.getNewRatings([1200, 900, 1000])
#  [1208.34629612,  910.43382278,  981.21988111]
```

See [`demo.md`](https://github.com/hoersamu/multi-elo/blob/main/demo.md) for a more in-depth tutorial, including details on parameters that can be tuned in the Elo algorithm.

## Methodology

For more Info on the methodology read [djcunningham0s Readme](https://github.com/djcunningham0/multielo/).

## Development

- Clone this repository
- Install latest LTS version of [Node.js](https://nodejs.org/en/)
- Enable [Corepack](https://github.com/nodejs/corepack) using `corepack enable`
- Install dependencies using `pnpm install`
- Run interactive tests using `pnpm dev`

## License

Made with 💛 by Samuel Höra

Published under [MIT License](./LICENSE).

<!-- Badges -->

[npm-version-src]: https://img.shields.io/npm/v/multi-elo?style=flat&colorA=18181B&colorB=F0DB4F
[npm-version-href]: https://npmjs.com/package/multi-elo
[npm-downloads-src]: https://img.shields.io/npm/dm/multi-elo?style=flat&colorA=18181B&colorB=F0DB4F
[npm-downloads-href]: https://npmjs.com/package/multi-elo
[codecov-src]: https://img.shields.io/codecov/c/gh/unjs/packageName/main?style=flat&colorA=18181B&colorB=F0DB4F
[codecov-href]: https://codecov.io/gh/unjs/packageName
[bundle-src]: https://img.shields.io/bundlephobia/minzip/multi-elo?style=flat&colorA=18181B&colorB=F0DB4F
[bundle-href]: https://bundlephobia.com/result?p=multi-elo


CHANGELOG.md
# Changelog


## v2.3.1

[compare changes](https://github.com/hoersamu/multi-elo/compare/v2.3.0...v2.3.1)

## v2.3.0


### 🚀 Enhancements

- Change to vitest, pnpm and update deps ([8d73760](https://github.com/hoersamu/multi-elo/commit/8d73760))

### 🩹 Fixes

- Empty lockfile ([e975e46](https://github.com/hoersamu/multi-elo/commit/e975e46))
- Add `coverage` directory to `.eslintignore` ([#62](https://github.com/hoersamu/multi-elo/pull/62))

### 📖 Documentation

- Fix yarn add command ([#21](https://github.com/hoersamu/multi-elo/pull/21))
- Add node lts info ([9af0642](https://github.com/hoersamu/multi-elo/commit/9af0642))
- Fix build status badge url ([#66](https://github.com/hoersamu/multi-elo/pull/66))

### 🏡 Chore

- Add development ([bc17492](https://github.com/hoersamu/multi-elo/commit/bc17492))
- Update script ([23448c8](https://github.com/hoersamu/multi-elo/commit/23448c8))
- Update tests and add coverage ([1eeb2a6](https://github.com/hoersamu/multi-elo/commit/1eeb2a6))
- Add `@vitest/coverage-c8` and run coverage in `test` script ([#13](https://github.com/hoersamu/multi-elo/pull/13))
- Use explicit dev dependency ([3f025c1](https://github.com/hoersamu/multi-elo/commit/3f025c1))
- Enable typescript strict mode ([#16](https://github.com/hoersamu/multi-elo/pull/16))
- Type exports ([#18](https://github.com/hoersamu/multi-elo/pull/18))
- Update renovate config ([#25](https://github.com/hoersamu/multi-elo/pull/25))
- Update template ([ef62c35](https://github.com/hoersamu/multi-elo/commit/ef62c35))
- Update deps ([5b82f23](https://github.com/hoersamu/multi-elo/commit/5b82f23))
- Update badges style and description ([9618507](https://github.com/hoersamu/multi-elo/commit/9618507))
- Update deps ([#82](https://github.com/hoersamu/multi-elo/pull/82))
- Add autofix ci ([03bae11](https://github.com/hoersamu/multi-elo/commit/03bae11))
- Remove lockfile from repo ([513a5e9](https://github.com/hoersamu/multi-elo/commit/513a5e9))
- Update template ([9d3de16](https://github.com/hoersamu/multi-elo/commit/9d3de16))
- Add bun install ([c675005](https://github.com/hoersamu/multi-elo/commit/c675005))
- Update dependencies ([58807bc](https://github.com/hoersamu/multi-elo/commit/58807bc))
- Remove `types` and `*build` from default gitignores ([ddf04f6](https://github.com/hoersamu/multi-elo/commit/ddf04f6))
- Simplify `pnpm run` to `pnpm` and update pnpm version ([#93](https://github.com/hoersamu/multi-elo/pull/93))

### ❤️ Contributors

- Samuel Höra <samuel.hoera@myposter.de>
- Nozomu Ikuta <nick.0508.nick@gmail.com>
- Pooya Parsa ([@pi0](http://github.com/pi0))
- Sébastien Chopin <seb@nuxtlabs.com>
- Daniel Roe <daniel@roe.dev>



src/utils.spec.ts
import { allClose, close, range } from "./utils";

describe("utils", () => {
  describe.each`
    size  | startAt
    ${5}  | ${0}
    ${10} | ${2}
    ${3}  | ${99}
    ${5}  | ${-5}
  `("range", ({ size, startAt }: { size: number; startAt: number }) => {
    it("should return correct ranges", () => {
      const calculatedRange = range(size, startAt);
      expect(calculatedRange.length).toEqual(size);
      for (let i = 0; i < size; i++) {
        expect(calculatedRange[i]).toEqual(startAt + i);
      }
    });
  });

  describe.each`
    a       | b           | maxDiff | expectedResult
    ${1}    | ${1}        | ${0}    | ${true}
    ${1}    | ${2}        | ${0}    | ${false}
    ${1}    | ${1 + 1e-6} | ${1e-5} | ${true}
    ${1000} | ${1016}     | ${1e-5} | ${false}
  `(
    "close",
    ({ a, b, maxDiff, expectedResult }: { a: number; b: number; maxDiff: number; expectedResult: boolean }) => {
      it("should validate max difference correctly", () => {
        expect(close(a, b, maxDiff)).toEqual(expectedResult);
      });
    },
  );

  describe.each`
    a                                               | b                                               | maxDiff | expected
    ${[1000, 1000]}                                 | ${[1000, 1000]}                                 | ${1e-5} | ${true}
    ${[1191.6880983472654, 1008.3119016527346]}     | ${[1191.6880983472654, 1008.3119016527346]}     | ${1e-5} | ${true}
    ${[1207.06479284, 989.33333333, 803.60187383]}  | ${[1207.06479284, 989.33333333, 803.60187383]}  | ${1e-5} | ${true}
    ${[1196.39812617, 1010.66666667, 792.93520716]} | ${[1196.39812617, 1010.66666667, 792.93520716]} | ${1e-5} | ${true}
    ${[1185.7314595, 1000, 814.2685405]}            | ${[1185.7314595, 1000, 814.2685405]}            | ${1e-5} | ${true}
    ${[1000, 1000]}                                 | ${[1000, 1001]}                                 | ${1e-5} | ${false}
    ${[1191.6880983472654, 1008.3119016527346]}     | ${[1191.6880983472654, 1008.4119016527346]}     | ${1e-5} | ${false}
    ${[1207.06479284, 989.33333333, 803.60187383]}  | ${[1207.06479284, 989.73333333, 803.60187383]}  | ${1e-5} | ${false}
    ${[1196.39812617, 1010.66666667, 792.93520716]} | ${[1196.39812617, 1010.06666667, 792.93520716]} | ${1e-5} | ${false}
    ${[1185.7314595, 1000, 814.2685405]}            | ${[1187.7314595, 1002, 816.2685405]}            | ${1e-5} | ${false}
  `("allClose", ({ a, b, maxDiff, expected }: { a: number[]; b: number[]; maxDiff: number; expected: boolean }) => {
    it("should validate max difference correctly", () => {
      expect(allClose(a, b, maxDiff)).toEqual(expected);
    });
  });
});


test/utils.spec.ts
import { expect, it, describe } from "vitest";
import { allClose, close, range } from "../src/utils";

describe("utils", () => {
  describe.each`
    size  | startAt
    ${5}  | ${0}
    ${10} | ${2}
    ${3}  | ${99}
    ${5}  | ${-5}
  `("range", ({ size, startAt }: { size: number; startAt: number }) => {
    it("should return correct ranges", () => {
      const calculatedRange = range(size, startAt);
      expect(calculatedRange.length).toEqual(size);
      for (let i = 0; i < size; i++) {
        expect(calculatedRange[i]).toEqual(startAt + i);
      }
    });
  });

  describe.each`
    a       | b           | maxDiff | expectedResult
    ${1}    | ${1}        | ${0}    | ${true}
    ${1}    | ${2}        | ${0}    | ${false}
    ${1}    | ${1 + 1e-6} | ${1e-5} | ${true}
    ${1000} | ${1016}     | ${1e-5} | ${false}
  `(
    "close",
    ({
      a,
      b,
      maxDiff,
      expectedResult,
    }: {
      a: number;
      b: number;
      maxDiff: number;
      expectedResult: boolean;
    }) => {
      it("should validate max difference correctly", () => {
        expect(close(a, b, maxDiff)).toEqual(expectedResult);
      });
    },
  );

  describe.each`
    a                                                     | b                                                     | maxDiff | expected
    ${[1000, 1000]}                                       | ${[1000, 1000]}                                       | ${1e-5} | ${true}
    ${[1191.688_098_347_265_4, 1008.311_901_652_734_6]}   | ${[1191.688_098_347_265_4, 1008.311_901_652_734_6]}   | ${1e-5} | ${true}
    ${[1207.064_792_84, 989.333_333_33, 803.601_873_83]}  | ${[1207.064_792_84, 989.333_333_33, 803.601_873_83]}  | ${1e-5} | ${true}
    ${[1196.398_126_17, 1010.666_666_67, 792.935_207_16]} | ${[1196.398_126_17, 1010.666_666_67, 792.935_207_16]} | ${1e-5} | ${true}
    ${[1185.731_459_5, 1000, 814.268_540_5]}              | ${[1185.731_459_5, 1000, 814.268_540_5]}              | ${1e-5} | ${true}
    ${[1000, 1000]}                                       | ${[1000, 1001]}                                       | ${1e-5} | ${false}
    ${[1191.688_098_347_265_4, 1008.311_901_652_734_6]}   | ${[1191.688_098_347_265_4, 1008.411_901_652_734_6]}   | ${1e-5} | ${false}
    ${[1207.064_792_84, 989.333_333_33, 803.601_873_83]}  | ${[1207.064_792_84, 989.733_333_33, 803.601_873_83]}  | ${1e-5} | ${false}
    ${[1196.398_126_17, 1010.666_666_67, 792.935_207_16]} | ${[1196.398_126_17, 1010.066_666_67, 792.935_207_16]} | ${1e-5} | ${false}
    ${[1185.731_459_5, 1000, 814.268_540_5]}              | ${[1187.731_459_5, 1002, 816.268_540_5]}              | ${1e-5} | ${false}
  `(
    "allClose",
    ({
      a,
      b,
      maxDiff,
      expected,
    }: {
      a: number[];
      b: number[];
      maxDiff: number;
      expected: boolean;
    }) => {
      it("should validate max difference correctly", () => {
        expect(allClose(a, b, maxDiff)).toEqual(expected);
      });
    },
  );

  it("should throw errors", () => {
    expect(() => allClose([1, 2], [1])).toThrowError(
      "a and b must have the same length",
    );
  });
});


src/multielo.spec.ts
import { MultiElo } from "./multielo";
import { sumReducer } from "./score-functions";
import { allClose, close } from "./utils";

describe("MultiElo", () => {
  describe.each`
    k     | d      | s       | ratings                    | trueExpected                                        | trueNew
    ${32} | ${400} | ${1}    | ${[1000, 1000]}            | ${[0.5, 0.5]}                                       | ${[1016, 984]}
    ${32} | ${400} | ${1}    | ${[1000, 1200]}            | ${[0.24025307, 0.75974693]}                         | ${[1024.31190165, 1175.68809835]}
    ${32} | ${400} | ${1}    | ${[1200, 800]}             | ${[0.90909091, 0.09090909]}                         | ${[1202.90909091, 797.09090909]}
    ${64} | ${400} | ${1}    | ${[1200, 1000]}            | ${[0.75974693, 0.24025307]}                         | ${[1215.37619669, 984.62380331]}
    ${64} | ${800} | ${1}    | ${[1200, 1000]}            | ${[0.640065, 0.359935]}                             | ${[1223.03584001, 976.96415999]}
    ${32} | ${800} | ${1}    | ${[1200, 1000]}            | ${[0.640065, 0.359935]}                             | ${[1211.51792001, 988.48207999]}
    ${32} | ${200} | ${1}    | ${[1200, 1000]}            | ${[0.90909091, 0.09090909]}                         | ${[1202.90909091, 997.09090909]}
    ${32} | ${400} | ${1.5}  | ${[1200, 1000]}            | ${[0.75974693, 0.24025307]}                         | ${[1207.68809835, 992.31190165]}
    ${32} | ${400} | ${1}    | ${[1200, 1000, 900]}       | ${[0.53625579, 0.29343936, 0.17030485]}             | ${[1208.34629612, 1002.55321444, 889.10048944]}
    ${32} | ${400} | ${1}    | ${[1000, 1200, 900]}       | ${[0.29343936, 0.53625579, 0.17030485]}             | ${[1023.88654777, 1187.01296279, 889.10048944]}
    ${32} | ${400} | ${1.25} | ${[1200, 1000, 900]}       | ${[0.53625579, 0.29343936, 0.17030485]}             | ${[1209.98732176, 1000.9121888, 889.10048944]}
    ${32} | ${400} | ${1.5}  | ${[1200, 1000, 900]}       | ${[0.53625579, 0.29343936, 0.17030485]}             | ${[1211.39391517, 999.50559539, 889.10048944]}
    ${32} | ${400} | ${2}    | ${[1200, 1000, 900]}       | ${[0.53625579, 0.29343936, 0.17030485]}             | ${[1213.67962945, 997.21988111, 889.10048944]}
    ${32} | ${400} | ${1.25} | ${[1200, 1000, 900, 1050]} | ${[0.38535873, 0.21814249, 0.13458826, 0.26191052]} | ${[1214.82857088, 1009.6423915, 900.67244749, 1024.85659012]}
  `(
    "test elo changes",
    ({
      k,
      d,
      s,
      ratings,
      trueExpected,
      trueNew,
    }: {
      k: number;
      d: number;
      s: number;
      ratings: number[];
      trueExpected: number[];
      trueNew: number[];
    }) => {
      const elo = new MultiElo({ k, d, s });
      it("should calculate correct expected scores", () => {
        expect(allClose(trueExpected, elo.getExpectedScores(ratings))).toBe(true);
      });
      it("should calculate correct new ratings", () => {
        expect(allClose(trueNew, elo.getNewRatings(ratings))).toBe(true);
      });
    },
  );

  describe.each([2, 3, 4, 10])(
    "make sure expected scores sum to 1 and rating changes are zero sum",
    (numPlayers: number) => {
      for (let i = 0; i < 10; i++) {
        const random = (min: number, max: number): number => {
          return Math.random() * (max - min) + min;
        };

        const k = random(16, 64);
        const d = random(200, 800);
        const ratings = Array(numPlayers).fill(random(600, 1400));
        const elo = new MultiElo({ k, d });

        it("expected scores should sum to 1", () => {
          expect(close(elo.getExpectedScores(ratings).reduce(sumReducer), 1)).toBe(true);
        });
        it("shoulnt change ratings sum", () => {
          expect(close(elo.getNewRatings(ratings).reduce(sumReducer), ratings.reduce(sumReducer))).toBe(true);
        });
      }
    },
  );

  describe.each`
    result               | resultOrder  | newRatings
    ${[1000, 1000]}      | ${[1, 1]}    | ${[1000, 1000]}
    ${[1200, 1000]}      | ${[0, 0]}    | ${[1191.6880983472654, 1008.3119016527346]}
    ${[1200, 1000, 800]} | ${[1, 2, 2]} | ${[1207.06479284, 989.33333333, 803.60187383]}
    ${[1200, 1000, 800]} | ${[1, 1, 2]} | ${[1196.39812617, 1010.66666667, 792.93520716]}
    ${[1200, 1000, 800]} | ${[1, 1, 1]} | ${[1185.7314595, 1000, 814.2685405]}
  `(
    "test ties",
    ({ result, resultOrder, newRatings }: { result: number[]; resultOrder: number[]; newRatings: number[] }) => {
      const elo = new MultiElo({ k: 32, d: 400 });
      it("should calculate correct tie scores", () => {
        expect(allClose(elo.getNewRatings(result, resultOrder), newRatings)).toBe(true);
      });
    },
  );

  describe.each`
    result               | resultOrder
    ${[1000, 1000]}      | ${[1, 0]}
    ${[1200, 1000]}      | ${[0, 0]}
    ${[1200, 1000, 800]} | ${[1, 2, 3]}
    ${[1200, 1000, 800]} | ${[1, 1, 2]}
  `("out of order ratings", ({ result, resultOrder }: { result: number[]; resultOrder: number[] }) => {
    const elo = new MultiElo({ k: 32, d: 400 });

    it("should match result in revered order", () => {
      expect(
        allClose(
          elo.getNewRatings(result, resultOrder),
          elo.getNewRatings(result.reverse(), resultOrder.reverse()).reverse(),
        ),
      ).toBe(true);
    });
  });
});


.eslintrc.js
/* eslint-disable @typescript-eslint/no-magic-numbers */
module.exports = {
  env: {
    browser: true,
    es6: true,
    node: true,
  },
  parser: "@typescript-eslint/parser",
  parserOptions: {
    project: ["./tsconfig.json"],
    sourceType: "module",
  },
  ignorePatterns: ["/lib/**/*", "**/*.html"],
  plugins: ["@typescript-eslint", "simple-import-sort"],
  overrides: [
    {
      files: ["*.ts", "*.tsx"], // Your TypeScript files extension

      // As mentioned in the comments, you should extend TypeScript plugins here,
      // instead of extending them outside the `overrides`.
      // If you don't want to extend any rules, you don't need an `extends` attribute.
      extends: [
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
      ],

      parserOptions: {
        project: ["./tsconfig.json"], // Specify it only for TypeScript files
      },
    },
  ],
  rules: {
    "no-restricted-syntax": [
      "warn",
      {
        selector: "CallExpression[callee.object.name='console'][callee.property.name!=/^(warn|error|info)$/]",
        message: "Do not use console.* for production. Allowed methods are console.info/warn/error.",
      },
    ],
    "@typescript-eslint/explicit-member-accessibility": [
      "error",
      {
        ignoredMethodNames: ["constructor"],
        overrides: {
          methods: "explicit",
          properties: "explicit",
          constructors: "explicit",
        },
      },
    ],
    "@typescript-eslint/indent": ["error", 2],
    "@typescript-eslint/member-delimiter-style": [
      "warn",
      {
        multiline: {
          delimiter: "semi",
          requireLast: true,
        },
        singleline: {
          delimiter: "semi",
          requireLast: false,
        },
      },
    ],
    "@typescript-eslint/explicit-function-return-type": "error",
    "@typescript-eslint/no-unsafe-assignment": "error",
    "@typescript-eslint/comma-spacing": ["warn"],
    "@typescript-eslint/typedef": [
      "error",
      {
        arrowParameter: false,
        variableDeclaration: false,
      },
    ],
    "@typescript-eslint/array-type": ["warn", { default: "array" }],
    "@typescript-eslint/no-misused-promises": "off",
    "@typescript-eslint/no-empty-function": "warn",
    "@typescript-eslint/no-misused-new": "warn",
    "@typescript-eslint/no-magic-numbers": [
      "warn",
      {
        ignore: [-1, 0, 1],
        ignoreArrayIndexes: true,
        detectObjects: true,
        ignoreNumericLiteralTypes: true,
        ignoreReadonlyClassProperties: true,
        ignoreEnums: true,
      },
    ],
    "@typescript-eslint/member-ordering": [
      "warn",
      {
        default: [
          "private-static-field",
          "private-instance-field",
          "protected-static-field",
          "protected-instance-field",
          "public-static-field",
          "public-instance-field",
          "private-constructor",
          "protected-constructor",
          "public-constructor",
          "method",
        ],
      },
    ],
    "@typescript-eslint/no-unused-vars": "off",
    "@typescript-eslint/no-useless-constructor": "off",
    "sort-imports": "off",
    "no-param-reassign": ["error", { props: false }],
    "brace-style": ["warn", "1tbs"],
    "@typescript-eslint/quotes": ["warn", "double"],
    "@typescript-eslint/semi": ["warn", "always"],
    "arrow-body-style": "warn",
    "comma-dangle": "off",
    "@typescript-eslint/no-inferrable-types": "off",
    complexity: [
      "warn",
      {
        max: 20,
      },
    ],
    "constructor-super": "warn",
    curly: "warn",
    eqeqeq: ["off", "always"],
    "id-blacklist": [
      "warn",
      "any",
      "Number",
      "number",
      "String",
      "string",
      "Boolean",
      "boolean",
      "Undefined",
      "undefined",
    ],
    "id-match": "warn",
    "max-classes-per-file": ["warn", 1],
    "no-cond-assign": "warn",
    "no-duplicate-case": "warn",
    "no-duplicate-imports": "warn",
    "no-empty": "warn",
    "no-invalid-this": "off",
    "no-irregular-whitespace": "warn",
    "no-multiple-empty-lines": "off",
    "no-new-wrappers": "warn",
    "no-redeclare": "warn",
    "no-shadow": "off",
    "@typescript-eslint/no-shadow": ["warn"],
    "no-template-curly-in-string": "warn",
    "no-throw-literal": "warn",
    "no-trailing-spaces": "warn",
    "no-underscore-dangle": "off",
    "no-unused-expressions": "error",
    "no-var": "warn",
    "one-var": ["warn", "never"],
    "prefer-const": "warn",
    "prefer-template": "warn",
    "quote-props": ["warn", "consistent"],
    radix: "warn",
    "space-before-function-paren": 0,
    "space-in-parens": 0,
    "spaced-comment": "warn",
    "use-isnan": "warn",
  },
};


src/utils.ts
/** set index to value */
const decor = (v: number, i: number): [number, number] => [v, i];
/** leave only index */
const undecor = (a: number[]): number => a[1];

export const argsort = (arr: number[]) =>
  arr
    .map((element, index) => decor(element, index))
    .sort()
    .map((element) => undecor(element));

/**
 * Returns an array of size 'size' filled with ascending numbers from starting with 'startAt'
 *
 * @param size the size of the generated array
 * @param startAt the starting index
 * @returns array filled with ascending numbers
 */
export function range(size: number, startAt = 0): number[] {
  const r = [...Array.from({ length: size }).keys()];
  if (startAt) {
    return r.map((i) => i + startAt);
  }
  return r;
}

/**
 * Checks if the absolute diff between a and b is smaller than maxDiff
 *
 * @param a first value
 * @param b second value
 * @param maxDiff maximum Difference (inclusive). Defaults to 1e-5
 */
export function close(a: number, b: number, maxDiff = 1e-5): boolean {
  return Math.abs(a - b) <= maxDiff;
}

/**
 * Checks if every value in array a is close to the value in array b
 *
 * @param a first array
 * @param b second array
 * @param maxDiff maximum Difference. Defaults to 1e-5
 */
export function allClose(a: number[], b: number[], maxDiff = 1e-5): boolean {
  if (a.length !== b.length) {
    throw new Error("a and b must have the same length");
  }
  for (const [index, element] of a.entries()) {
    if (!close(element, b[index], maxDiff)) {
      return false;
    }
  }
  return true;
}


src/setupPackage.ts
import * as fs from "fs";

// DO NOT DELETE THIS FILE
// This file is used by build system to build a clean npm package with the compiled js files in the root of the package.
// It will not be included in the npm package.

function main() {
  const source = fs.readFileSync(__dirname + "/../package.json").toString("utf-8");
  const sourceObj = JSON.parse(source);
  sourceObj.scripts = {};
  sourceObj.devDependencies = {};
  if (sourceObj.main.startsWith("dist/")) {
    sourceObj.main = sourceObj.main.slice(5);
  }
  fs.writeFileSync(__dirname + "/package.json", Buffer.from(JSON.stringify(sourceObj, null, 2), "utf-8"));

  fs.copyFileSync(__dirname + "/../.npmignore", __dirname + "/.npmignore");
  fs.copyFileSync(__dirname + "/../README.md", __dirname + "/README.md");
  fs.copyFileSync(__dirname + "/../LICENSE", __dirname + "/LICENSE");
}

main();


src/score-functions.ts
import { range } from "./utils";

export const sumReducer = (previousValue: number, currentValue: number) =>
  previousValue + currentValue;

export type ScoreFunction = (n: number) => number[];

/**
 * With the linear score function the "points" awarded scale linearly from first place
 * through last place. For example, improving from 2nd to 1st place has the same sized
 * benefit as improving from 5th to 4th place.
 *
 * @param n number of players
 * @returns array of the points to assign to each place (summing to 1)
 */
const linearScoreFunction = (n: number): number[] => {
  return range(n, 1).map((p) => (n - p) / ((n * (n - 1)) / 2));
};

/**
 * With an exponential score function with base > 1, more points are awarded to the top
 * finishers and the point distribution is flatter at the bottom. For example, improving
 * from 2nd to 1st place is more valuable than improving from 5th place to 4th place. A
 * larger base value means the scores will be more weighted towards the top finishers.
 *
 * @param n number of players
 * @param base base for the exponential score function (> 1)
 * @returns array of the points to assign to each place (summing to 1)
 */
const exponentialScoreFunction = (n: number, base: number): number[] => {
  if (base < 1) {
    throw new Error("Base must be >= 1");
  }

  const out: number[] = range(n, 1).map((p) => Math.pow(base, n - p) - 1);
  // eslint-disable-next-line unicorn/no-array-reduce
  const sum = out.reduce((accumulator, element) =>
    sumReducer(accumulator, element),
  );

  return out.map((p) => p / sum);
};

export const createExponentialScoreFuntion = (base: number): ScoreFunction => {
  return base === 1
    ? linearScoreFunction
    : (n: number) => exponentialScoreFunction(n, base);
};


src/multielo.ts
import {
  createExponentialScoreFuntion,
  ScoreFunction,
  sumReducer,
} from "./score-functions";
import { argsort, close, range } from "./utils";

export const DEFAULT_LOG_BASE_VALUE = 10;
export const DEFAULT_D_VALUE = 400;
export const DEFAULT_K_VALUE = 32;
export const DEFAULT_S_VALUE = 1;

export type MultiEloConfig = {
  logBase: number;
  d: number;
  k: number;
  s: number;
  verbose: boolean;
};

export class MultiElo {
  private static instance: MultiElo;
  private scoreFunction: ScoreFunction;
  private console: Console | { log: () => void };
  private config: MultiEloConfig = {
    logBase: DEFAULT_LOG_BASE_VALUE,
    d: DEFAULT_D_VALUE,
    k: DEFAULT_K_VALUE,
    s: DEFAULT_S_VALUE,
    verbose: false,
  };

  constructor(config: Partial<MultiEloConfig> = {}) {
    Object.assign(this.config, config);
    this.scoreFunction = createExponentialScoreFuntion(this.config.s);
    this.console = this.config.verbose
      ? console
      : { log: (): void => undefined };
  }

  /**
   * Update ratings based on results. Takes an array of ratings before the matchup and returns an array with
   * the updated ratings. Provided array should be ordered by the actual results (first place finisher's
   * initial rating first, second place next, and so on).
   *
   * Example usage:
   * >>> elo.getNewRatings([1200, 1000])
   * [1207.68809835,  992.31190165]
   * >>> elo.getNewRatings([1200, 1000, 1100, 900])
   * [1212.01868209, 1012.15595083, 1087.84404917,  887.98131791]
   *
   * @param initialRatings array of ratings (float values) in order of actual results
   * @param resultOrder array where each value indicates the place the player in the same index of
   *                    initialRatings finished in. Lower is better. Identify ties by entering the same value for players
   *                    that tied. For example, [1, 2, 3] indicates that the first listed player won, the second listed player
   *                    finished 2nd, and the third listed player finished 3rd. [1, 2, 2] would indicate that the second
   *                    and third players tied for 2nd place. Defaults to [1,2,3...]
   * @returns array of updated ratings in same order as input
   */
  getNewRatings(initialRatings: number[], resultOrder?: number[]) {
    const n = initialRatings.length;
    resultOrder = resultOrder || range(n, 0);
    const actualScores = this.getActualScores(n, resultOrder);
    const expectedScores = this.getExpectedScores(initialRatings);
    const scaleFactor = this.config.k * (n - 1);
    this.console.log("scaleFactor: " + scaleFactor);

    return initialRatings.map(
      (m, i) => m + scaleFactor * (actualScores[i] - expectedScores[i]),
    );
  }

  /**
   * Return the score to be awarded to the players based on the results.
   *
   * @param n number of players in the matchup
   * @param resultOrder list indicating order of finish. See getNewRatings docstring for more info
   * @return array of length n of scores to be assigned to firstplace, second place, and so on
   */
  private getActualScores(n: number, resultOrder?: number[]): number[] {
    // calculate actual scores according to score function, then sort in order of finish
    resultOrder = resultOrder || range(n, 0);
    let scores = this.scoreFunction(n);
    scores = argsort(argsort(resultOrder)).map((i) => scores[i]);

    // if there are ties, average the scores of all tied players
    const distinctResults = new Set(resultOrder);
    if (distinctResults.size !== n) {
      for (const place of distinctResults) {
        // eslint-disable-next-line unicorn/no-array-reduce
        const idx = resultOrder.reduce((accumulator, value, index) => {
          if (value === +place) {
            accumulator.push(index);
          }
          return accumulator;
        }, [] as number[]);

        const mean =
          idx
            .map((i) => scores[i])
            // eslint-disable-next-line unicorn/no-array-reduce
            .reduce((accumulator, element) =>
              sumReducer(accumulator, element),
            ) / idx.length;
        for (const i of idx) {
          scores[i] = mean;
        }
      }
    }

    this.console.log("resultOrder:", resultOrder);
    this.console.log("Calculated actual scores:", scores);
    this.validateActualScores(scores, resultOrder);
    return scores;
  }

  /**
   * Helper function for getActualScores. Runs multiple checks to validate the plausiblity of the new scores
   *
   * @param scores list of all player scores
   * @param resultOrder list indicating order of finish. See getNewRatings docstring for more info
   */
  private validateActualScores(scores: number[], resultOrder: number[]) {
    // eslint-disable-next-line unicorn/no-array-reduce
    const scoreSum = scores.reduce((accumulator, element) =>
      sumReducer(accumulator, element),
    );
    if (!close(1, scoreSum)) {
      throw new Error("Scoring function does not return scores summing to 1");
    }

    if (Math.min(...scores) !== 0) {
      // tie for last place means minimum score doesn't have to be zero,
      // so only raise error if there isn't a tie for last
      const lastPlace = Math.max(...resultOrder);
      if (resultOrder.filter((f) => f === lastPlace).length === 1) {
        throw new Error("Scoring function does not return minimum value of 0");
      }
    }

    // TODO: implement check to make sure scores are monotonically decreasing
  }

  /**
   * Get the expected scores for all players given their ratings before the matchup.
   *
   * @param initialRatings array of ratings for each player in a matchup
   * @returns array of expected scores for all players
   */
  getExpectedScores(ratings: number[]): number[] {
    this.console.log("Calculating expected scores for: ", ratings);

    // get all pairwise differences
    const diffMx = ratings.map((m) => ratings.map((mn) => mn - m));
    this.console.log("DiffMx: ", diffMx);

    // get individual contributions to expected score using logistic function
    const logisticMx = diffMx.map((m, i) =>
      m.map((mn, j) =>
        i === j
          ? 0
          : 1 / (1 + Math.pow(this.config.logBase, mn / this.config.d)),
      ),
    );
    this.console.log("LogisticMx: ", logisticMx);

    const n = ratings.length;
    // number of individual head-to-head matchups between n players
    const denom = (n * (n - 1)) / 2;
    const expectedScores = logisticMx.map(
      (m) =>
        // eslint-disable-next-line unicorn/no-array-reduce
        m.reduce((accumulator, element) => sumReducer(accumulator, element)) /
        denom,
    );

    // this should be guaranteed, but check to make sure
    if (
      !close(
        1,
        // eslint-disable-next-line unicorn/no-array-reduce
        expectedScores.reduce((accumulator, element) =>
          sumReducer(accumulator, element),
        ),
      )
    ) {
      throw new Error("Expected Scores do not sum to 1");
    }

    this.console.log("Calculated expected scores: ", expectedScores);

    return expectedScores;
  }

  static getNewRatings(oldRatings: number[], resultOrder?: number[]) {
    return this.getInstance().getNewRatings(oldRatings, resultOrder);
  }

  static getExpectedScores(ratings: number[]): number[] {
    return this.getInstance().getExpectedScores(ratings);
  }

  private static getInstance() {
    if (!this.instance) {
      this.instance = new MultiElo();
    }
    return this.instance;
  }
}

export function getNewRatings(
  oldRatings: number[],
  resultOrder?: number[],
): number[] {
  return MultiElo.getNewRatings(oldRatings, resultOrder);
}

export function getExpectedScores(ratings: number[]): number[] {
  return MultiElo.getExpectedScores(ratings);
}
